File: sdk/typescript/src/sandbox/lifecycle.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import { SandboxTimeoutError, SandboxValidationError } from "../errors/index.js";
import type { SandboxMetadata, SandboxOptions, SandboxState } from "./index.js";

export const defaultSandboxArgv = ["/bin/sh", "-lc", "sleep infinity"];

const runStatus = {
  running: 3,
  succeeded: 4,
  failed: 5,
  cancelled: 6,
} as const;

export function validateSandboxOptions(options: SandboxOptions): void {
  const sourceCount = [options.image, options.environmentId].filter(
    (source) => source !== undefined && source !== "",
  ).length;
  if (sourceCount !== 1) {
    throw new SandboxValidationError("exactly one of image or environmentId is required");
  }
}

export function sandboxLabels(labels: Record<string, string> | undefined): Record<string, string> {
  return {
    "axern.sdk": "typescript",
    ...(labels ?? {}),
  };
}

export function sandboxMetadata(options: SandboxOptions, state: SandboxState): SandboxMetadata {
  return {
    ...state,
    namespace: options.namespace ?? "default",
    labels: sandboxLabels(options.labels),
    source: sandboxSource(options),
  };
}

export async function waitRunningRun(
  runId: string,
  timeoutMs: number,
  watchRun: (runId: string, signal: AbortSignal) => AsyncIterable<Record<string, unknown>>,
): Promise<Record<string, unknown>> {
  const deadline = Date.now() + timeoutMs;
  let lastRun: Record<string, unknown> | undefined;
  const controller = new AbortController();
  const iterator = watchRun(runId, controller.signal)[Symbol.asyncIterator]();
  try {
    for (;;) {
      const remainingMs = deadline - Date.now();
      if (remainingMs <= 0) break;
      const result = await nextBeforeDeadline(iterator, remainingMs);
      if (result === undefined || result.done) break;
      const run = result.value;
      lastRun = run;
      const status = Number(run.status ?? 0);
      if (status === runStatus.running && String(run.allocation_id ?? "") !== "") {
        return run;
      }
      if (status === runStatus.succeeded || status === runStatus.failed || status === runStatus.cancelled) {
        throw new Error(`run ${runId} became ${String(run.status ?? "")} before its sandbox allocation was running: ${String(run.message ?? "")}`);
      }
    }
  } finally {
    controller.abort();
    try {
      const close = iterator.return?.();
      if (close !== undefined) void close.catch(() => undefined);
    } catch {
      // The readiness result owns the error surface; stream cleanup is best effort.
    }
  }
  const details = lastRun === undefined ? "no state observed" : `${String(lastRun.status ?? "")}: ${String(lastRun.message ?? "")}`;
  throw new SandboxTimeoutError(`run ${runId} did not reach a running sandbox allocation: ${details}`);
}

async function nextBeforeDeadline<T>(
  iterator: AsyncIterator<T>,
  timeoutMs: number,
): Promise<IteratorResult<T> | undefined> {
  let timer: number | undefined;
  try {
    return await Promise.race([
      iterator.next(),
      new Promise<undefined>((resolve) => {
        timer = setTimeout(resolve, timeoutMs);
      }),
    ]);
  } finally {
    if (timer !== undefined) clearTimeout(timer);
  }
}

function sandboxSource(options: SandboxOptions): SandboxMetadata["source"] {
  if (options.image !== undefined && options.image !== "") {
    return "image";
  }
  return "environment";
}

File: sdk/python/src/axern_sdk/sandbox/lifecycle.py
"""Run lifecycle helpers for SDK sandboxes."""

from __future__ import annotations

from contextlib import closing
import time

import grpc

from axern.control.run.v1 import run_pb2
from axern_sdk.client import AxernClient
from axern_sdk.errors import SandboxTimeoutError


def wait_running_run(
    client: AxernClient,
    *,
    run_id: str,
    timeout_seconds: float,
) -> run_pb2.Run:
    deadline = time.monotonic() + timeout_seconds
    last_run: run_pb2.Run | None = None
    try:
        with closing(client.watch_run(run_id, timeout=timeout_seconds)) as watch:
            for run in watch:
                last_run = run
                if run.status == run_pb2.RUN_STATUS_RUNNING and run.allocation_id:
                    return run
                if run.status in {run_pb2.RUN_STATUS_SUCCEEDED, run_pb2.RUN_STATUS_FAILED, run_pb2.RUN_STATUS_CANCELLED}:
                    raise RuntimeError(
                        f"run {run_id} became {run_pb2.RunStatus.Name(run.status)} "
                        f"before its sandbox allocation was running: {run.message}"
                    )
    except TimeoutError:
        pass
    except grpc.RpcError as exc:
        if exc.code() != grpc.StatusCode.DEADLINE_EXCEEDED or time.monotonic() < deadline:
            raise
    details = "no state observed" if last_run is None else f"{run_pb2.RunStatus.Name(last_run.status)}: {last_run.message}"
    raise SandboxTimeoutError(
        f"run {run_id} did not reach a running sandbox allocation within {timeout_seconds}s: {details}"
    )
